### ISSUE TITLE
$*DISTRO version not reporting accurate version for Windows running virtually

### ISSUE THREAD

**Issue by sdondley** [rakudo/rakudo#5042]
## The Problem

`say $*DISTRO` is reporting:

`mswin32 (10.0.19042.746)`

on a machine with Windows 11 installed. 

## Expected Behavior

As reported by the `ver` command on the guest Windows machine, the actual version is:

`Microsoft Windows [Version 10.0.22000.856]`

And it is a 64 bit install, not mswin32.

## Environment
* Operating system: Windows 11

* Compiler version (`perl6 -v` or `raku -v`):
2022.07

**Comment by sdondley** [comment]
Raku was installed with rakubrew. I suspect that rakubrew is deriving the os version from the OS version used in Strawberry Perl.

To see if this might be right, I downloaded a rakudo binary and installed it. Using this installed binary, Raku now reports the os version differently: 10.0.17763.529.

**Comment by lizmat** [comment]
What does
````raku
$*VM.config<osname>
````
say ?

**Comment by ugexe** [comment]
I think you're just mistaken that Windows 11 is windows version 11. See https://en.wikipedia.org/wiki/Windows_11_version_history#Version_22H2

**Comment by sdondley** [comment]
It says "MSWin32" (despite this being a 64 bit install of Windows 11).

**Comment by lizmat** [comment]
Somewhere in the Rakudo configuration process, that string is picked up or assumed.  Whether that is a problem of the configuration script, or just Windows reporting something old to indicate compatibility, I have no idea.  My last Windows machine was 20 years ago now, so...

**Comment by sdondley** [comment]
Yeah, I wondered the same, thing, if this old version might be some kind of compatibility issue. 

But I searched through Window's registry of the machine and found no partial match to the version raku reported. But it did match exactly what was in Strawberry Perl. So I gotta imagine rakubrew picked it from there.

And the fact that the rakudo binary reports a totally different version for windows than the raku version built with rakubrew leads me to believe it has to have to something to with the version of perl used in the process. That sure would be weird, though.

**Comment by Kaiepi** [comment]
On MoarVM, I think the `osname` value in `Distro`'s config maps to [`$^O`](https://perldoc.perl.org/variables/$%5EO) for the platform its `Configure.pl` was run on, which always reports `MSWin32` on Windows. This script generates `src/gen/config.c`, containing `MVM_backend_config`, which backs the `nqp::backendconfig` op.

**Comment by sdondley** [comment]
I'm not very qualified to make recommendations for such low level code, but it seems like the file at https://github.com/rakudo/rakudo/blob/1fc8cce31a370149326e029bd51eb884b7610205/src/core.c/Distro.pm6 should execute some commands similar with what's there for macOS to determine the OS version of a machine running windows.

**Comment by lizmat** [comment]
Well... I would be open to implement that, if someone could come up with the magic incantations for (all/most versions of) Windows.

**Comment by sdondley** [comment]
Judging by https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions, Looks like there are probably less than 20. I wouldn't worry about anything before say 2012 (when Windows 8 came out). A developer running anything older isn't likely to take Raku for a spin.

**Comment by sdondley** [comment]
And the `ver` command is what is issued from the cmd shell, which should work on all versions of Windows. That results in something like this: `Microsoft Windows [Version 10.0.22000.856]`

**Comment by sdondley** [comment]
See also: https://www.windows-commandline.com/find-windows-os-version-from-command/

The `systeminfo` command has a ton of info:

```
C:\Users\sdond>systeminfo

Host Name:                 WIN11
OS Name:                   Microsoft Windows 11 Pro
OS Version:                10.0.22000 N/A Build 22000
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Standalone Workstation
OS Build Type:             Multiprocessor Free
Registered Owner:          sdondley@gmail.com
Registered Organization:
Product ID:                00330-80000-00000-AA419
Original Install Date:     8/23/2022, 3:31:50 PM
System Boot Time:          8/24/2022, 7:49:03 AM
System Manufacturer:       innotek GmbH
System Model:              VirtualBox
System Type:               x64-based PC
Processor(s):              1 Processor(s) Installed.
                           [01]: Intel64 Family 6 Model 165 Stepping 5 GenuineIntel ~3600 Mhz
BIOS Version:              innotek GmbH VirtualBox, 12/1/2006
Windows Directory:         C:\Windows
System Directory:          C:\Windows\system32
Boot Device:               \Device\HarddiskVolume1
System Locale:             en-us;English (United States)
Input Locale:              en-us;English (United States)
Time Zone:                 (UTC-05:00) Eastern Time (US & Canada)
Total Physical Memory:     16,371 MB
Available Physical Memory: 10,976 MB
Virtual Memory: Max Size:  19,315 MB
Virtual Memory: Available: 13,587 MB
Virtual Memory: In Use:    5,728 MB
Page File Location(s):     C:\pagefile.sys
Domain:                    WORKGROUP
Logon Server:              \\WIN11
Hotfix(s):                 6 Hotfix(s) Installed.
                           [01]: KB5015732
                           [02]: KB5010690
                           [03]: KB5012170
                           [04]: KB5016629
                           [05]: KB5006755
                           [06]: KB5015898
Network Card(s):           1 NIC(s) Installed.
                           [01]: Intel(R) PRO/1000 MT Desktop Adapter
                                 Connection Name: Ethernet
                                 DHCP Enabled:    Yes
                                 DHCP Server:     10.0.2.2
                                 IP address(es)
                                 [01]: 10.0.2.15
                                 [02]: fe80::5988:c70a:bac1:5c83
Hyper-V Requirements:      A hypervisor has been detected. Features required for Hyper-V will not be displayed.
```

**Comment by jonathanstowe** [comment]
Windows appears to have an API for that (similarly I haven't written software for windows in nearly twenty years,) but it seems rather heavily caveated. https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getversion

**Comment by vrurg** [comment]
> Judging by https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions, looks like there are probably less than 20 if you don't worry about anything before say 2012 (when Windows 8 came out). A developer running anything older isn't likely to take Raku for a spin.

Surprise! For work purposes I have a Win7 VM and it's the same VM I use to test Windows builds when necessary. I barely expect anything newer to be used.

**Comment by sdondley** [comment]
Looks like this Perl module can save a lot of wheel reinventing: https://metacpan.org/pod/Win32#Win32::GetOSName()

**Comment by ugexe** [comment]
This might be relevant:

win: support Windows 11 in uv_os_uname() - https://github.com/libuv/libuv/commit/97dcdb1926f6aca43171e1614338bcef067abd59
